Add `ConcurrencyLimit` and its default implementation

trustin opened the pull request:
Motivation:

As a preparatory step for resolving #12842, we need a construct that limits concurrency asynchronously.

Modifications:

- Add `ConcurrencyLimit` and `ConcurrencyLimitHandler`
- Add the default implementation `SimpleConcurrencyLimit`

Result:

- In our future work, we can allow a user to specify a `ConcurrencyLimit` when building a `DnsNameResolver` and use it for limiting the concurrency of DNS queries in `DnsQueryContext.writeQuery()`

normanmaurer commented:
@trustin welcome back buddy :)

franz1981 reviewed (changes requested):
I would allow using different queues types for pendingHandlers and would save the PendingHandler in a local variable outside of the CAS loop to reuse it in case of failing CAS.

Furthermore, not a blocker for this, I would implement a more constrained version (with lighter concurrent ops) for single threaded usage too (where it makes sense, to be used always by the same event loop)

hyperxpro commented on common/src/test/java/io/netty/util/concurrent/SimpleConcurrencyLimitTest.java:
```diff
@@ -0,0 +1,154 @@
+                    assertEquals(expectedPermitsInUseOnAcquisition, limit.permitsInUse());
+                    releaser.run();
+                    assertEquals(expectedPermitsInUseOnAcquisition - 1, limit.permitsInUse());
+                    events.add("acquired");
```
I think it'd be good if we declare class-level variables for event names.

chrisvest commented on common/src/main/java/io/netty/util/concurrent/ConcurrencyLimitHandler.java:
```diff
@@ -0,0 +1,38 @@
+     *
+     * @param releaser the {@link Runnable} that releases the permit
+     */
+    void permitAcquired(Runnable releaser);
```
What's the idea behind relying on the callee to invoke the releaser, vs. automatically releasing a permit upon the completion of this method?

chrisvest commented on common/src/main/java/io/netty/util/concurrent/SimpleConcurrencyLimit.java:
```diff
@@ -0,0 +1,197 @@
+                    pendingHandler.setTimeoutFuture(timeoutFuture);
+                }
+
+                pendingHandlers.add(pendingHandler);
+                return;
```
We have a data-race here where all permits could have been returned, and all releasers could have been called, before we manage to add our pending handler to the queue.

chrisvest commented on common/src/main/java/io/netty/util/concurrent/SimpleConcurrencyLimit.java:
```diff
@@ -0,0 +1,197 @@
+        for (;;) {
+            final int currentPermitsInUse = permitsInUse;
+            if (currentPermitsInUse == maxPermits) {
+                pendingHandlers.add(pendingHandler);
+                return;
```
Same data-race exists here, with a smaller window.

chrisvest commented on common/src/main/java/io/netty/util/concurrent/SimpleConcurrencyLimit.java:
```diff
@@ -0,0 +1,197 @@
+                assert nextPermitInUse >= 0 : "Released more than acquired";
+                if (permitsInUseUpdater.compareAndSet(SimpleConcurrencyLimit.this,
+                                                      currentPermitInUse, nextPermitInUse)) {
+                    final PendingHandler pendingHandler = pendingHandlers.poll();
```
If the pending handler is non-null, we could in theory "hand over" our existing permit, and skip the decrement/increment dance.

chrisvest reviewed (commented):
Some comments

trustin commented:
I renamed `ConcurrencyLimit` to `Limiter` and `SimpleConcurrencyLimiter` to `ConcurrencyLimiter`, and then added `RateLimiter`. I didn't address the comments yet and there are still some rough edges in the implementation, so I converted this PR to draft. I'll ping you once it's ready for reviews.

trustin commented:
Had a chat with @normanmaurer a few days ago and we concluded that at the moment we don't really need any pluggability for rate-limiting DNS queries - just simple rate limiting implemented directly in our resolver implementation will do. Once we find similar use cases in other places in Netty, we may resurrect this PR, but for now, let me close it.
